beacon-chain/core/validators/validator.go

```diff
@@ -179,11 +179,70 @@ func ActivateValidator(state *pb.BeaconState, idx uint64, genesis bool) (*pb.Bea
 //
 // Spec pseudocode definition:
 // def initiate_validator_exit(state: BeaconState, index: ValidatorIndex) -> None:
+//    """
+//    Initiate the validator of the given ``index``.
+//    Note that this function mutates ``state``.
+//    """
+//    # Return if validator already initiated exit
 //    validator = state.validator_registry[index]
-//    validator.status_flags |= INITIATED_EXIT
+//    if validator.exit_epoch != FAR_FUTURE_EPOCH:
+//        return
+//
+//    # Compute exit queue epoch
+//    exit_epochs = [v.exit_epoch for v in state.validator_registry if v.exit_epoch != FAR_FUTURE_EPOCH]
+//    exit_queue_epoch = max(exit_epochs + [get_delayed_activation_exit_epoch(get_current_epoch(state))])
+//    exit_queue_churn = len([v for v in state.validator_registry if v.exit_epoch == exit_queue_epoch])
+//    if exit_queue_churn >= get_churn_limit(state):
+//        exit_queue_epoch += 1
+//
+//    # Set validator exit epoch and withdrawable epoch
+//    validator.exit_epoch = exit_queue_epoch
+//    validator.withdrawable_epoch = validator.exit_epoch + MIN_VALIDATOR_WITHDRAWABILITY_DELAY
 func InitiateValidatorExit(state *pb.BeaconState, idx uint64) *pb.BeaconState {
-	state.ValidatorRegistry[idx].StatusFlags |=
-		pb.Validator_INITIATED_EXIT
+	v := state.ValidatorRegistry[idx]
+
+	// Return if validator already initiated exit.
+	// According to the spec, this is not an assert condition and
+	// shouldn't fail beacon block state transition.
+	if v.ExitEpoch != params.BeaconConfig().FarFutureEpoch {
+		return state
+	}
+
+	// Aggregate all the exit epochs from the last.
+	var exitedEpochs []uint64
```

Comment:
It looks like you are looking only for the highest epoch, no need to store all of them in an array and iterate over them to find the highest epoch.

Reply:
great feedback, this simplified it by a lot